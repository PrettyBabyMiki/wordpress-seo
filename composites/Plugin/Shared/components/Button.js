import React from "react";
import styled from "styled-components";
import flow from "lodash/flow";
import omit from "lodash/omit";
import PropTypes from "prop-types";

import colors from "../../../../style-guide/colors.json";
import { Icon } from "./Icon";
import { rgba } from "../../../../style-guide/helpers";

const settings = {
	minHeight: 32,
	verticalPadding: 4,
	borderWidth: 1,
};

const ieMinHeight = settings.minHeight - ( settings.verticalPadding * 2 ) - ( settings.borderWidth * 2 );

/**
 * Returns a component with applied base button styles.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied base button styles.
 */
export function addBaseStyle( component ) {
	return styled( component )`
		display: inline-flex;
		align-items: center;
		justify-content: center;
		vertical-align: middle;
		border-width: ${ `${ settings.borderWidth }px` };
		border-style: solid;
		margin: 0;
		padding: ${ `${ settings.verticalPadding }px` } 10px;
		border-radius: 3px;
		cursor: pointer;
		box-sizing: border-box;
		font-size: inherit;
		font-family: inherit;
		font-weight: inherit;
		text-align: left;
		overflow: visible;
		min-height: ${ `${ settings.minHeight }px` };

		svg {
			// Safari 10
			align-self: center;
		}

		// Only needed for IE 10+.
		@media all and ( -ms-high-contrast: none ), ( -ms-high-contrast: active ) {
			::after {
				display: inline-block;
				content: "";
				min-height: ${ `${ ieMinHeight }px` };
			}
		}
	`;
}

/**
 * Returns a component with applied focus styles.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied focus styles.
 */
export function addFocusStyle( component ) {
	return styled( component )`
		&::-moz-focus-inner {
			border-width: 0;
		}

		&:focus {
			outline: none;
			border-color: ${ colors.$color_blue };
			background-color: ${ colors.$color_white };
			box-shadow: 0 0 3px ${ rgba( colors.$color_blue_dark, .8 ) };
		}
	`;
}

/**
 * Returns a component with applied hover styles.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied hover styles.
 */
export function addHoverStyle( component ) {
	return styled( component )`
		&:hover {
			background: ${ colors.$color_button_hover };
			border-color: ${ colors.$color_button_border_hover };
			color: ${ colors.$color_button_text_hover };
		}
	`;
}

/**
 * Returns a component with applied active styles.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied active styles.
 */
export function addActiveStyle( component ) {
	return styled( component )`
		&:active {
			background: ${ colors.$color_button };
			border-color: ${ colors.$color_button_border_hover };
			box-shadow: inset 0 2px 5px -3px ${ rgba( colors.$color_button_border_active, 0.5 ) };
		}
	`;
}

/**
 * Returns a component with applied font size style.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied font size styles.
 */
export function addFontSizeStyles( component ) {
	return styled( component )`
		font-size: 0.8rem;
	`;
}

/**
 * Returns a component with all button selector styles applied.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied styles.
 */
export const addButtonStyles = _flow( [ addBaseStyle, addFocusStyle, addHoverStyle, addActiveStyle ] );

/**
 * Returns a basic styled button.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} Styled button.
 */
export const BaseButton = addButtonStyles(
	styled.button`
		color: ${ props => props.textColor };
		border-color: ${ props => props.borderColor };
		background: ${ props => props.backgroundColor };
		box-shadow: 0 1px 0 ${ props => rgba( props.boxShadowColor, 1 ) };
	`
);

BaseButton.propTypes = {
	backgroundColor: PropTypes.string,
	textColor: PropTypes.string,
	borderColor: PropTypes.string,
	boxShadowColor: PropTypes.string,
};

BaseButton.defaultProps = {
	backgroundColor: colors.$color_button,
	textColor: colors.$color_button_text,
	borderColor: colors.$color_button_border,
	boxShadowColor: colors.$color_button_border,
	type: "button",
};

/**
 * Returns a styled Button with set font size.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} Styled button.
 */
export const Button = addFontSizeStyles( BaseButton );

/**
 * Applies styles to icon for IconButton with text.
 *
 * @param {ReactElement} icon The original icon.
 *
 * @returns {ReactElement} Icon with text styles applied.
 */
function addIconTextStyle( icon ) {
	return styled( icon )`
		margin: 0 8px 0 0;
		flex-shrink: 0;
	`;
}

/**
 * Returns an icon button that can optionally contain text.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} Styled icon button.
 */
export const IconButton = ( props ) => {
	const { children: text, icon, iconColor } = props;

	let IconComponent = Icon;
	if ( text ) {
		IconComponent = addIconTextStyle( IconComponent );
	}

	const newProps = omit( props, "icon" );

	return (
		<Button { ...newProps }>
			<IconComponent icon={ icon } color={ iconColor } />
			{ text }
		</Button>
	);
};

IconButton.propTypes = {
	icon: PropTypes.func.isRequired,
	iconColor: PropTypes.string.isRequired,
	children: PropTypes.string,
};
