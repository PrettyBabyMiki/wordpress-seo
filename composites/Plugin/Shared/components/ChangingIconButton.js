import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";

import colors from "../../../../style-guide/colors.json";
import { Icon } from "./Icon";
import { rgba } from "../../../../style-guide/helpers";

export const ChangingIconButtonBase = styled.button`
	box-sizing: border-box;
	width: 32px;
	display: inline-block;
	border: 1px solid ${ colors.$color_button_border };
	background-color: ${ props => props.pressed ? props.pressedBackground : props.unpressedBackground };
	box-shadow: ${ props => props.pressed
		? `inset 0 2px 0 ${ rgba( props.pressedBoxShadowColor, 0.7 ) }`
		: `0 1px 0 ${ rgba( props.unpressedBoxShadowColor, 0.7 ) }` };
	border-radius: 3px;
	cursor: pointer;
	outline: none;
	height: ${ props => props.pressed ? "24px" : "25px" };

	/* 
	Each button has an icon and a secondary icon to make it possible to switch icons on hover.
	The secondary icon is initially not displayed.
	*/
	.yoast-changing-icon-button__secondary-icon {
			display: none;
	}
	
	// Switch icons on hover.
	&:hover .yoast-changing-icon-button__icon {
		display: none;
	}

	&:hover .yoast-changing-icon-button__secondary-icon {
		display: inline;
	}
`;

const ChangingIconButton = ( props ) => {
	return (
		<ChangingIconButtonBase
			onClick={ props.onClick }
			pressed={ props.pressed }
			unpressedBoxShadowColor={ props.unpressedBoxShadowColor }
			pressedBoxShadowColor={ props.pressedBoxShadowColor }
			pressedBackground={ props.pressedBackground }
			unpressedBackground={ props.unpressedBackground }
			aria-label={ props.id }
			aria-pressed={ props.pressed }
			unpressedIconColor={ props.unpressedIconColor }
			pressedIconColor={ props.pressedIconColor }
		>
			{ props.pressed
				? <Icon className="yoast-changing-icon-button__icon"
					icon={ props.pressedIcon }
					color={ props.pressedIconColor }
					size="18px"/>
				: <Icon className="yoast-changing-icon-button__icon"
					icon={ props.unpressedIcon }
					color={ props.unpressedIconColor }
					size="18px"/>
			}
			{ props.pressed
				? <Icon className="yoast-changing-icon-button__secondary-icon"
					icon={ props.unpressedIcon }
					color={ props.pressedIconColor }
					size="18px"/>
				: <Icon className="yoast-changing-icon-button__secondary-icon"
					icon={ props.pressedIcon }
					color={ props.unpressedIconColor }
					size="18px"/>
			}
		</ChangingIconButtonBase>
	);
};

ChangingIconButton.propTypes = {
	id: PropTypes.string.isRequired,
	onClick: PropTypes.func.isRequired,
	boxShadowColor: PropTypes.string,
	unpressedBoxShadowColor: PropTypes.string,
	pressedBoxShadowColor: PropTypes.string,
	pressedBackground: PropTypes.string,
	unpressedBackground: PropTypes.string,
	pressedIconColor: PropTypes.string,
	unpressedIconColor: PropTypes.string,
	pressedIcon: PropTypes.func.isRequired,
	unpressedIcon: PropTypes.func.isRequired,
	pressed: PropTypes.bool.isRequired,
};

ChangingIconButton.defaultProps = {
	unpressedBoxShadowColor: colors.$color_button_border,
	pressedBoxShadowColor: colors.$color_purple,
	pressedBackground: colors.$color_pink_dark,
	unpressedBackground: colors.$color_button,
	pressedIconColor: colors.$color_white,
	unpressedIconColor: colors.$color_button_text,
};

export default ChangingIconButton;
