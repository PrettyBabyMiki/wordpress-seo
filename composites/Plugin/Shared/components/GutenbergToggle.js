/**
 * External dependencies
 */
import classnames from "classnames";
import noop from "lodash/noop";

function FormToggle( { className, checked, id, onChange = noop, ...props } ) {
	const wrapperClasses = classnames(
		'components-form-toggle',
		className,
		{ 'is-checked': checked },
	);

	return (
		<span className={ wrapperClasses }>
			<input
				className="components-form-toggle__input"
				id={ id }
				type="checkbox"
				checked={ checked }
				onChange={ onChange }
				{ ...props }
			/>
			<span className="components-form-toggle__track"></span>
			<span className="components-form-toggle__thumb"></span>
			{ checked ?
				<svg className="components-form-toggle__on" width="2" height="6" aria-hidden="true" role="img" focusable="false" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 2 6"><path d="M0 0h2v6H0z" /></svg> : <svg className="components-form-toggle__off" width="6" height="6" aria-hidden="true" role="img" focusable="false" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 6 6"><path d="M3 1.5c.8 0 1.5.7 1.5 1.5S3.8 4.5 3 4.5 1.5 3.8 1.5 3 2.2 1.5 3 1.5M3 0C1.3 0 0 1.3 0 3s1.3 3 3 3 3-1.3 3-3-1.3-3-3-3z" /></svg>
			}
		</span>
	);
}

export default FormToggle;
