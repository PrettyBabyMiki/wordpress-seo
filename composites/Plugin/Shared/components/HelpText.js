/* External dependencies */
import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const YoastHelpText = styled.p`
	font-size: 13px;
`;

/**
 * Returns the HelpText component.
 *
 * @param {Object} props Component props.
 *
 * @returns {ReactElement} HelpText component.
 */
export default class HelpText extends React.Component {
	/**
	 * Renders a help text component.
	 *
	 * @returns {ReactElement} The rendered help text component.
	 */
	render() {
		const { text } = this.props;

		return (
			<YoastHelpText>
				{ text }
			</YoastHelpText>
		);
	}
}

/**
 * React prop type for the help text.
 *
 * Use this in your components to pass along the text.
 */
export const HelpTextPropType = PropTypes.oneOfType( [
	PropTypes.string,
	PropTypes.array,
] );

HelpText.propTypes = {
	text: HelpTextPropType.isRequired,
};
