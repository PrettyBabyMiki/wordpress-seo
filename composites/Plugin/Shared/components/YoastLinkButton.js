import PropTypes from "prop-types";
import styled from "styled-components";
import colors from "../../../../style-guide/colors.json";
import { rgba } from "../../../../style-guide/helpers";
import { addButtonStyles } from "./YoastButton";

/**
 * Returns a Button with the Yoast button style.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} Styled button.
 */
export const YoastLinkButton = addButtonStyles(
	styled.a`
		text-decoration: none;
		color: ${ props => props.textColor };
		background: ${ props => props.backgroundColor };
		min-width: 152px;
		${ props => props.withTextShadow ? `text-shadow: 0 0 2px ${ colors.$color_black }` : "" };
	`
);

YoastLinkButton.propTypes = {
	backgroundColor: PropTypes.string,
	textColor: PropTypes.string,
	withTextShadow: PropTypes.bool,
};

YoastLinkButton.defaultProps = {
	backgroundColor: colors.$color_green_medium_light,
	textColor: colors.$color_white,
	withTextShadow: true,
};
