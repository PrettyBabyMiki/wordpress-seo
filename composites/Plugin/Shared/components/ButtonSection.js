import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";

import colors from "../../../../style-guide/colors.json";
import { IconsButton } from "../../Shared/components/Button";
import ScreenReaderText from "../../../../a11y/ScreenReaderText";
import { getRtlStyle } from "../../../../utils/helpers/styled-components";

const StyledContainer = styled.div`
	background-color: ${ colors.$color_white };
`;

const StyledContainerTopLevel = styled( StyledContainer )`
	border-top: 1px solid ${ colors.$palette_grey_light };
	border-bottom: 1px solid ${ colors.$palette_grey_light };
	margin-top: -1px;
`;

export const StyledIconsButton = styled( IconsButton )`
	width: 100%;
	background-color: ${ colors.$color_white };
	padding: 16px;
	justify-content: flex-start;
	border-color: transparent;
	border: none;
	border-radius: 0;
	box-shadow: none;

	:focus {
		outline: 1px solid ${ colors.$color_blue };
		outline-offset: -1px;
	}

	:active {
		box-shadow: none;
		background-color: ${ colors.$color_white };
	}

	svg {
		${ props => props.hasSubTitle ? "align-self: flex-start;" : "" }
		&:first-child {
			${ getRtlStyle( "margin-right: 8px", "margin-left: 8px" ) };
		}
		&:last-child {
			${ getRtlStyle( "margin-left: 8px", "margin-right: 8px" ) };
		}
	}
`;

const StyledTitleContainer = styled.span`
	flex-grow: 1;
	overflow-x: hidden;
	line-height: normal; // Avoid vertical scrollbar in IE 11 when rendered in the WP sidebar.
`;

const StyledTitle = styled.span`
	display: block;
	white-space: nowrap;
	text-overflow: ellipsis;
	overflow-x: hidden;
`;

const StyledSubTitle = styled.span`
	display: block;
	white-space: nowrap;
	text-overflow: ellipsis;
	overflow-x: hidden;
	font-size: 0.8125rem;
	margin-top: 2px;
`;

/**
 * Wraps a component in a heading element with a defined heading level.
 *
 * @param {ReactElement} Component        The component to wrap.
 * @param {Object}       props            The heading props.
 * @param {number}       props.level      The heading level.
 * @param {string}       props.fontSize   The heading font-size.
 * @param {string}       props.fontWeight The heading font-weight.
 *
 * @returns {Function} A function that will return the wrapped component with given properties.
 */
export function wrapInHeading( Component, props ) {
	const Heading = `h${ props.level }`;
	const StyledHeading = styled( Heading )`
		margin: 0 !important;
		padding: 0 !important;
		font-size: ${ props.fontSize } !important;
		font-weight: ${ props.fontWeight } !important;
	`;

	return function Wrapped( props ) {
		return (
			<StyledHeading>
				<Component { ...props } />
			</StyledHeading>
		);
	};
}

const StyledHeading = wrapInHeading( StyledIconsButton, { level: 2, fontSize: "1rem", fontWeight: "normal" } );

/**
 * Base panel. Optionally has a heading around the button.
 *
 * @param {Object}      props                       The properties for the component.
 * @param {string}      props.className             The name of the CSS class used for alternate styling.
 * @param {IconsButton} props.Heading               Heading button. May be wrapped or styled or both.
 * @param {boolean}     props.hasSeparator          True displays borders around the section. False means no borders.
 * @param {Object}      props.prefixIcon            Heading icon before the title.
 * @param {string}      props.subTitle              Sub-title for the Heading.
 * @param {Object}      props.suffixIcon            Heading icon after the title.
 * @param {string}      props.title                 Title for the Heading.
 * @param {string}      props.titleScreenReaderText Chance for an extra text to feed to a screenreader.
 *
 * @returns {ReactElement} A panel that can be clicked to open a modal.
 */
export const ButtonSection = ( props ) => {
	const Container = ( props.hasSeparator ) ? StyledContainerTopLevel : StyledContainer;

	return (
		<Container
			// Pass the classname to allow re-styling with styled-components.
			className={ props.className }
		>
			<props.Heading
				prefixIcon={ props.prefixIcon }
				suffixIcon={ props.suffixIcon }
				hasSubTitle={ !! props.subTitle }
			>
				<StyledTitleContainer>
					<StyledTitle>
						{ props.title }
						{ props.titleScreenReaderText && <ScreenReaderText>{ " " + props.titleScreenReaderText }</ScreenReaderText> }
					</StyledTitle>
					{ props.subTitle && <StyledSubTitle>{ props.subTitle }</StyledSubTitle> }
				</StyledTitleContainer>
			</props.Heading>
		</Container>
	);
};

ButtonSection.propTypes = {
	className: PropTypes.string,
	Heading: PropTypes.func,
	isOpen: PropTypes.bool.isRequired,
	hasSeparator: PropTypes.bool,
	hasPadding: PropTypes.bool,
	prefixIcon: PropTypes.shape( {
		icon: PropTypes.string,
		color: PropTypes.string,
		size: PropTypes.string,
	} ),
	subTitle: PropTypes.string,
	suffixIcon: PropTypes.shape( {
		icon: PropTypes.string,
		color: PropTypes.string,
		size: PropTypes.string,
	} ),
	title: PropTypes.string.isRequired,
	titleScreenReaderText: PropTypes.string,
};

ButtonSection.defaultProps = {
	Heading: StyledHeading,
};

export default ButtonSection;
