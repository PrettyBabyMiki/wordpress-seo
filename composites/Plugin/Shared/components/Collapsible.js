import PropTypes from "prop-types";
import React from "react";
import styled from "styled-components";
import omit from "lodash/omit";

import colors from "../../../../style-guide/colors.json";
import { IconsButton } from "../../Shared/components/Button";
import ScreenReaderText from "../../../../a11y/ScreenReaderText";

const StyledContainer = styled.div`
	background-color: ${ colors.$color_white };
	border-top: 1px solid ${ colors.$color_grey };
	border-bottom: 1px solid ${ colors.$color_grey };
	margin-top: -1px;
`;

const StyledContent = styled.div`
	padding: 0 15px 15px;

	& p:last-child {
		margin-bottom: 0;
	}
`;

const StyledIconsButton = styled( IconsButton )`
	width: 100%;
	background-color: ${ colors.$color_white };
	padding: 15px;
	justify-content: flex-start;
	border-color: transparent;
	border-radius: 0;
	box-shadow: none;
	color: ${ colors.$color_button_border_active };

	svg {
		&:first-child {
			margin-right: 8px;
		}
		&:last-child {
			margin-left: 8px;
		}
	}
`;

const StyledTitle = styled.span`
	white-space: nowrap;
	text-overflow: ellipsis;
	overflow-x: hidden;
	-ms-flex-positive: 1;
	flex-grow: 1;
	font-size: 1.25rem;
`;

/**
 * Wraps a component in a heading element with a defined heading level.
 *
 * @param {ReactElement} Component    The component to wrap.
 * @param {int}          headingLevel The heading level.
 *
 * @returns {Function} A function that will return the wrapped component with given properties.
 */
export function wrapInHeading( Component, headingLevel ) {
	const Heading = `h${ headingLevel }`;
	const StyledHeading = styled( Heading )`
		margin: 0;
		font-weight: normal;
	`;

	return function Wrapped( props ) {
		return (
			<StyledHeading>
				<Component { ...props } />
			</StyledHeading>
		);
	};
}

const StyledHeading = wrapInHeading( StyledIconsButton, 2 );

/**
 * Base collapsible panel. Optionally has a heading around the button.
 *
 * @param {object}      props                       The properties for the component.
 * @param {children}    props.children              The content of the Collapsible.
 * @param {IconsButton} props.Heading               Heading button. May be wrapped or styled or both.
 * @param {boolean}     props.isOpen                True displays the children. False means collapsed.
 * @param {function}    props.onToggle              Function to handle the Heading click event.
 * @param {string}      props.prefixIcon            Heading icon before the title.
 * @param {string}      props.prefixIconCollapsed   Prefix icon when in collapsed state.
 * @param {string}      props.prefixIconColor       CSS color of the prefix icon.
 * @param {string}      props.suffixIcon            Heading icon after the title.
 * @param {string}      props.suffixIconColor       CSS color of the suffix icon.
 * @param {string}      props.suffixIconCollapsed   Suffix icon when in collapsed state.
 * @param {string}      props.prefixIconViewBox     The viewBox for the prefix SVG icon element.
 * @param {string}      props.suffixIconViewBox     The viewBox for the suffix SVG icon element.
 * @param {string}      props.title                 Title for in the Heading.
 * @param {string}      props.titleScreenReaderText Chance for an extra text to feed to a screenreader.
 *
 * @returns {ReactElement} A collapsible panel.
 */
export const CollapsibleStateless = ( props ) => {
	return (
		<StyledContainer>
			<props.Heading
				aria-expanded={ props.isOpen }
				onClick={ props.onToggle }
				prefixIcon={ props.isOpen ? props.prefixIcon : props.prefixIconCollapsed }
				prefixIconColor={ props.prefixIconColor }
				suffixIcon={ props.isOpen ? props.suffixIcon : props.suffixIconCollapsed }
				suffixIconColor={ props.suffixIconColor }
				prefixIconViewBox={ props.prefixIconViewBox }
				suffixIconViewBox={ props.suffixIconViewBox }
				prefixIconSize={ props.prefixIconSize }
				suffixIconSize={ props.suffixIconSize }
			>
				<StyledTitle>{ props.title }</StyledTitle>
				{ props.titleScreenReaderText ? <ScreenReaderText>{ " " + props.titleScreenReaderText }</ScreenReaderText> : null }
			</props.Heading>
			{ props.isOpen && props.children }
		</StyledContainer>
	);
};

CollapsibleStateless.propTypes = {
	children: PropTypes.oneOfType( [
		PropTypes.arrayOf( PropTypes.node ),
		PropTypes.node,
	] ),
	Heading: PropTypes.func,
	isOpen: PropTypes.bool.isRequired,
	onToggle: PropTypes.func.isRequired,
	prefixIcon: PropTypes.string,
	prefixIconCollapsed: PropTypes.string,
	prefixIconColor: PropTypes.string,
	prefixIconSize: PropTypes.string,
	prefixIconViewBox: PropTypes.string,
	suffixIcon: PropTypes.string,
	suffixIconCollapsed: PropTypes.string,
	suffixIconColor: PropTypes.string,
	suffixIconSize: PropTypes.string,
	suffixIconViewBox: PropTypes.string,
	title: PropTypes.string.isRequired,
	titleScreenReaderText: PropTypes.string,
};

CollapsibleStateless.defaultProps = {
	Heading: StyledHeading,
	prefixIconColor: colors.$black,
	suffixIconColor: colors.$black,
};

/**
 * Stateful collapsible panel. Optionally has a heading around the button.
 */
export class Collapsible extends React.Component {
	/**
	 * The constructor.
	 *
	 * @param {object}  props                       The properties for the component.
	 * @param {number}  props.headingLevel          Heading level: 1 for h1, 2 for h2, etc.
	 * @param {boolean} props.initialIsOpen         Determines if the initial isOpen state is open or closed.
	 * @param {string}  props.prefixIcon            Heading icon before the title.
	 * @param {string}  props.prefixIconCollapsed   Prefix icon when in collapsed state.
	 * @param {string}  props.prefixIconColor       CSS color of the prefix icon.
	 * @param {string}  props.suffixIcon            Heading icon after the title.
	 * @param {string}  props.suffixIconColor       CSS color of the suffix icon.
	 * @param {string}  props.suffixIconCollapsed   Suffix icon when in collapsed state.
	 * @param {string}  props.title                 Title for in the Heading.
	 * @param {string}  props.titleScreenReaderText Chance for an extra text to feed to a screenreader.
	 *
	 * @returns {ReactElement} Base collapsible panel.
	 */
	constructor( props ) {
		super( props );

		this.state = {
			isOpen: props.initialIsOpen,
		};

		/*
		 * Evaluate if the button should be wrapped in a heading in this constructor
		 * instead of doing it in the render function to avoid a full re-render of the button,
		 * which is bad for accessibility.
		 */
		this.Heading = Collapsible.getHeading( props );
		this.toggleCollapse = this.toggleCollapse.bind( this );
	}

	/**
	 * Makes sure the heading element is correctly set.
	 *
	 * @param {Object} nextProps The upcoming props.
	 *
	 * @returns {void}
	 */
	componentWillReceiveProps( nextProps ) {
		const { headingLevel } = this.props;

		if ( nextProps.headingLevel !== headingLevel ) {
			this.Heading = Collapsible.getHeading( nextProps );
		}
	}

	/**
	 * Toggles whether the list is collapsed.
	 *
	 * @returns {void}
	 */
	toggleCollapse() {
		const { isOpen } = this.state;

		this.setState( {
			isOpen: ! isOpen,
		} );
	}

	/**
	 * Creates the header by wrapping the IconsButton with a header.
	 *
	 * @param {object} props The properties for the component.
	 *
	 * @returns {ReactElement} The header to render.
	 */
	static getHeading( props ) {
		return wrapInHeading( StyledIconsButton, props.headingLevel );
	}

	/**
	 * Returns the rendered collapsible panel.
	 *
	 * @returns {ReactElement} The rendered collapsible panel.
	 */
	render() {
		const { isOpen } = this.state;
		const { children, prefixIconViewBox, suffixIconViewBox } = this.props;

		const newProps = omit( this.props, [ "children" ] );

		return (
			<CollapsibleStateless
				Heading={ this.Heading }
				isOpen={ isOpen }
				onToggle={ this.toggleCollapse }
				{ ...newProps }
				prefixIconViewBox={ prefixIconViewBox }
				suffixIconViewBox={ suffixIconViewBox }
			>
				{ isOpen && <StyledContent>{ children }</StyledContent> }
			</CollapsibleStateless>
		);
	}
}

Collapsible.propTypes = {
	children: PropTypes.oneOfType( [
		PropTypes.arrayOf( PropTypes.node ),
		PropTypes.node,
	] ),
	headingLevel: PropTypes.number,
	initialIsOpen: PropTypes.bool,
	prefixIcon: PropTypes.string,
	prefixIconCollapsed: PropTypes.string,
	prefixIconColor: PropTypes.string,
	prefixIconViewBox: PropTypes.string,
	suffixIcon: PropTypes.string,
	suffixIconCollapsed: PropTypes.string,
	suffixIconColor: PropTypes.string,
	suffixIconViewBox: PropTypes.string,
	title: PropTypes.string.isRequired,
	titleScreenReaderText: PropTypes.string,
};

Collapsible.defaultProps = {
	headingLevel: 2,
	initialIsOpen: false,
	prefixIconColor: colors.$black,
	suffixIcon: "arrow-up",
	suffixIconCollapsed: "arrow-down",
	suffixIconColor: colors.$black,
};

export default Collapsible;
