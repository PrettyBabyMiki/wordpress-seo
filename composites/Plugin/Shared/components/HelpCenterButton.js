import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import colors from "../../../../style-guide/colors.json";
import { questionCircle, angleUp, angleDown } from "../../../../style-guide/svg";
import { YoastButton } from "./YoastButton";
import { Icon } from "../../Shared/components/Icon";

/**
 * Returns an icon button that can optionally contain text.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} Styled icon button.
 */
const HelpCenterButtonBase = ( props ) => {
	return (
		<YoastButton { ...props }>
			<Icon icon={ questionCircle } color={ props.iconColor } />
			{ props.children }
			<Icon icon={ props.isExpanded ? angleUp : angleDown } color={ props.iconColor } />
		</YoastButton>
	);
};

export const HelpCenterButton = styled( HelpCenterButtonBase )`
	min-width: 0;
	padding: 0 2px;

	svg {
		margin: 0 16px;
	}
`;

HelpCenterButtonBase.propTypes = {
	iconColor: PropTypes.string,
	isExpanded: PropTypes.bool,
	children: PropTypes.oneOfType( [
		PropTypes.arrayOf( PropTypes.node ),
		PropTypes.node,
		PropTypes.string,
	] ),
};

HelpCenterButtonBase.defaultProps = {
	iconColor: colors.$color_white,
	isExpanded: false,
};
