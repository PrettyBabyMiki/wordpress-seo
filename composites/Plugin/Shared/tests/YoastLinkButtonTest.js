import React from "react";
import renderer from "react-test-renderer";

import { YoastLinkButton } from "../components/YoastLinkButton";

test( "the YoastLinkButton matches the snapshot", () => {
	const component = renderer.create(
		<YoastLinkButton>YoastLinkButtonValue</YoastLinkButton>
	);

	let tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );
