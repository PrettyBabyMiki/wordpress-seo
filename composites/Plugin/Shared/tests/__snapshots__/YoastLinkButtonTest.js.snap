// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the YoastLinkButton matches the snapshot 1`] = `
<a
  className="sc-bxivhb kzimrn sc-htpNat czwPQj"
>
  YoastLinkButtonValue
</a>
`;
